language/move-model/src/ty.rs

```diff
@@ -618,38 +621,39 @@ impl Substitution {
             (Type::Tuple(ts1), Type::Tuple(ts2)) => {
                 return Ok(Type::Tuple(self.unify_vec(
                     display_context,
+                    variance,
                     ts1,
                     ts2,
                     "tuples",
                 )?));
             }
             (Type::Fun(ts1, r1), Type::Fun(ts2, r2)) => {
                 return Ok(Type::Fun(
-                    self.unify_vec(display_context, ts1, ts2, "functions")?,
-                    Box::new(self.unify(display_context, Variance::Disallow, &*r1, &*r2)?),
+                    self.unify_vec(display_context, variance, ts1, ts2, "functions")?,
+                    Box::new(self.unify(display_context, variance, &*r1, &*r2)?),
                 ));
             }
             (Type::Struct(m1, s1, ts1), Type::Struct(m2, s2, ts2)) => {
                 if m1 == m2 && s1 == s2 {
                     return Ok(Type::Struct(
                         *m1,
                         *s1,
-                        self.unify_vec(display_context, ts1, ts2, "structs")?,
+                        self.unify_vec(display_context, variance, ts1, ts2, "structs")?,
                     ));
                 }
             }
             (Type::Vector(e1), Type::Vector(e2)) => {
                 return Ok(Type::Vector(Box::new(self.unify(
                     display_context,
-                    Variance::Disallow,
```

Comment:
That we disallow covariance on type arguments is actually an intentional design. Variance in the current implementation is only allowed on the outmost level of a type unification, but not if you descent down into types.

This is a common restriction you find in most languages which support both coercion (integer types) and parametric polymorphism. The reason is that if you, say, look at `Fun<I, O>`, you don't know whether I and O are [co- or contra-variant](https://en.wikipedia.org/wiki/Covariance_and_contravariance_(computer_science). 

Perhaps our type system doesn't require the distinction of co and contra, because we don't really have subtyping, but just normalization to num and removal of addresses. But this would need to be shown first. As said, most languages (for sure Java, I think also C++, and perhaps Rust) don't allow variance in type parameters,

The concrete reason why we have this restriction is actually in our compilation scheme. Lets say we have compiled to a `vector<u64>`. This type is not compatible with `vector<num>`, that is the variance is not reflected in the monimorphized compilation result. That is why I think I introduced this restriction (hitting real bugs). We would have to normalized all types ahead of unification time instead of using variance to solve this problem, which would be a lot of work.

Reply:
Thanks for the detailed explanation Wolfgang! Now I see why variance should be only applicable to the outer-most level in type unification.

I think my confusion mainly comes from what this `variance` parameter means here. In our implementation, it means all integer types, i.e., `u8`, `u64`, `u128`, and `num` are compatible. And if in our Move code we write sth that has type `vector<u8>` and in our spec we inferred sth of type `vector<u128>` and try to compare these two vectors, we probably should allow the comparison? An example is:

```
module 0x42::M {
  use Std::Vector;

  fun foo(): vector<u8> {
    let v = Vector::empty();
    Vector::push_back(&mut v, 1);
    v
  }

  spec foo {
    ensures result == vec(1);
  }
}
```

Before this PR, we run into error
```
+ error: no matching declaration of `==`
+
+     ┌── tests/sources/functional/num_type_unification.move:11:13 ───
+     │
+  11 │     ensures result == vec(1);
+     │             ^^^^^^^^^^^^^^^^
+     │
+     = outruled candidate `==(#0, #0): bool` (expected `u8` but found `u128` for argument 2)
```

After this PR, we could verify this test case.

Reply:
Its controlling the types in function arguments and also in struct constructors, because the type checker calls unify on each single parameter/ctor argument. Variance is a standard notion for what we are doing here (its not just ints but also refs), and its also pretty standard to _not_ propagate it over type constructors.

Reply:
FWIW the non-propagation was added together with monomorphization as required by that compilation scheme. The code here is more general. We can change the enum variants to reflect this (e.g. have a general variance which propagates, have a shallow only, and none at all) but I would leave the general logic here.

Reply:
> We can change the enum variants to reflect this (e.g. have a general variance which propagates, have a shallow only, and none at all) but I would leave the general logic here.

The PR is updated in this direction.